Dispatch params/pdoc 37

iankronquist opened the pull request:
@hkenney

hkenney commented on lib/puppet_x/puppetlabs/strings/yard/handlers/puppet_4x_function_handler.rb:
```diff
@@ -23,6 +23,25 @@ class Puppet4xFunctionHandler < YARD::Handlers::Ruby::Base
+    param_name_ident = type_specifier.jump :ident
+    return [] if param_name_ident == type_specifier
+    param_name = param_name_ident.source
+          #require 'pry'; binding.pry
```
;)

hkenney commented on lib/puppet_x/puppetlabs/strings/yard/handlers/puppet_4x_function_handler.rb:
```diff
@@ -23,6 +23,25 @@ class Puppet4xFunctionHandler < YARD::Handlers::Ruby::Base
+    return [] if param_name_ident == type_specifier
+    param_name = param_name_ident.source
+          #require 'pry'; binding.pry
+    #puts [param_name, param_signature].inspect
```
Commented out debug lines ;)

iankronquist replied:
I should write a post commit hook which greps for these. Thanks.

hkenney commented:
Except the debug statements I am :+1: on this. Really awesome work @iankronquist nice job.

@hlindberg Could you take a second look at this? As someone who's familiar with Puppet AST nodes and the parser.

hlindberg commented on lib/puppet_x/puppetlabs/strings/yard/handlers/puppet_4x_function_handler.rb:
```diff
@@ -33,22 +52,73 @@ class Puppet4xFunctionHandler < YARD::Handlers::Ruby::Base
+    end
+
+    # If the overload_signatures list is empty because we couldn't find any
+    # dispatch blocks, then there must be one function named the same as the
```
"must be one method" instead of "must be one function" (would make the comment more clear).

hlindberg commented on lib/puppet_x/puppetlabs/strings/yard/handlers/puppet_4x_function_handler.rb:
```diff
@@ -33,22 +52,73 @@ class Puppet4xFunctionHandler < YARD::Handlers::Ruby::Base
+            if param.type == :ident
+              # The parameters of Puppet functions with no defined dispatch are
+              # as though they are Any type.
+              signature << [param[0], 'Any']
```
A ruby parameter may be a splat (e.g. def foo(a, b, *c) which we take as variable number of arguments

hlindberg commented on lib/puppet_x/puppetlabs/strings/yard/handlers/puppet_4x_function_handler.rb:
```diff
@@ -33,22 +52,73 @@ class Puppet4xFunctionHandler < YARD::Handlers::Ruby::Base
+    obj.type_info = overload_signatures
+
+    # The yard docstring parser expects a list of lists, not a list of lists of
+    # lists.
```
(picky) - in Ruby "list" is referred to as "array"

hlindberg commented on spec/unit/puppet_x/puppetlabs/strings/yard/puppet_4x_function_handler_spec.rb:
```diff
@@ -71,4 +71,74 @@ def max(num_a, num_b)
   end
+
+
+  it "should not issue a warning if the parameter names do  match the docstring" do
```
Better to phrase test "names" as a heading for an example rather than expressing a requirements ("should") - e.g. "does not issue a warning when parameter names match", or "no warning is issued when parameter names match".

hlindberg commented on spec/unit/puppet_x/puppetlabs/strings/yard/puppet_4x_function_handler_spec.rb:
```diff
@@ -71,4 +71,74 @@ def max(num_a, num_b)
+
+
+  it "should issue a warning if the parameter names do not match the " +
+        "docstring in dispatch method" do
```
avoid using '+' in the title

hlindberg commented:
I would like to see an example that uses parameterized types e.g. Integer[2,10]

iankronquist commented:
@hlindberg 
1. I updated the comments. I like pedantic PRs, they're so easy to fix.
2. I did not update the test header string to remove "should". All of the other tests in the file begin with "should" and it is my understanding that RSpec tests should begin this way. Perhaps I did not quite understand your issue with my phrasing, so I revised it so hopefully it is clearer.
3. I added tests which use parameterized types for both the positive and negative cases.

hkenney commented:
@hlindberg Let me know when you're happy with this, I'll give it a final once over and merge

hlindberg commented:
+1
None of the examples seems to use a parameterized type (should work, but nice to have a test for that)
Need a ticket for handling 'captures rest' (varargs) functionality.

iankronquist commented:
@hlindberg sorry about that. I revised two of the tests (a positive and a negative case) to reflect that. The tests pass locally and on travis so I squashed and believe this is ready to merge.
